lib/preload.js

```diff
@@ -1,32 +1,125 @@
 'use strict'
 
-const childProcess = require('child_process')
 const path = require('path')
-const escape = require('command-join')
-const compareVersions = require('compare-versions')
+const Docker = require('dockerode')
+const tarfs = require('tar-fs')
+const stream = require('stream')
+const Promise = require('bluebird')
 
 const preload = module.exports
 
+const docker = new Docker({Promise: Promise})
+
+const DOCKER_IMAGE_TAG = 'resin/resin-preload'
+const DISK_IMAGE_PATH_IN_DOCKER = '/img/resin.img'
+const SPLASH_IMAGE_PATH_IN_DOCKER = '/img/resin-logo.png'
+
 /** @const {String} Default container name */
 preload.CONTAINER_NAME = 'resin-image-preloader'
 
 /**
  * Build the preloader docker image
- * @param {Object} [processOptions] - child process options
- * @returns {ChildProcess}
+ * @returns Promise
  */
-preload.build = function (processOptions) {
-  processOptions = Object.assign({
-    cwd: process.cwd(),
-    env: process.env,
-    stdio: 'pipe',
-    detached: false
-  }, processOptions)
-
-  const dirname = path.join(__dirname, '..')
-  const argv = [ 'build', '-t', 'resin/resin-preload', dirname ]
-
-  return childProcess.spawn('docker', argv, processOptions)
+preload.build = function () {
+  console.log('Building preloader image...')
+  return docker.buildImage(
+    tarfs.pack(path.resolve(__dirname, '..')),
+    {t: DOCKER_IMAGE_TAG}
+  )
+  .then((stream_) => {
+    return new Promise((resolve, reject) => {
+      docker.modem.followProgress(
+        stream_,
+        (err, output) => {
+          if (err) {
+            reject(err)
+            return
+          }
+          resolve()
+        },
+        (event) => {
+          if (event.stream) {
+            process.stdout.write(event.stream)
+          }
+        }
+      )
+    })
+  })
+}
+
+class DevNull extends stream.Writable {
+  _write (chunk, enc, next) {
+    next()
+  }
+}
+
+class BufferBackedWritableStream extends stream.Writable {
+  constructor () {
+    super(arguments)
+    this.chunks = []
+  }
+
+  _write (chunk, enc, next) {
+    this.chunks.push(chunk)
+    next()
+  }
+
+  getData () {
+    return Buffer.concat(this.chunks)
+  }
+}
+
+const bindMount = (source, target) => {
+  return {
+    Source: path.resolve(source),
+    Target: target,
+    Type: 'bind',
+    Consistency: 'delegated'
+  }
+}
+
+const createContainerDisposer = (options) => {
```

Comment:
Really nice idea :+1: I feel like the builder could benefit from this kind of interface!